v2/tools/mangle-test-json/main.go

```diff
@@ -0,0 +1,251 @@
+				panic("run while already running")
+			}
+			startTimes[key(d)] = d.Time
+		case "pause":
+			if startTimes[key(d)] == (time.Time{}) {
+				panic("pause while not running")
+			}
+			runTimes[key(d)] += d.Time.Sub(startTimes[key(d)])
+			startTimes[key(d)] = time.Time{}
+		case "cont":
+			// cont while still in running state happens sometimes (???)
+			// so don't check
+			startTimes[key(d)] = d.Time
+		case "output":
+			outputs[key(d)] = append(outputs[key(d)], d.Output)
+		case "pass", "fail", "skip":
+			if d.Test != "" && startTimes[key(d)] == (time.Time{}) {
+				panic("finished when not running")
+			}
+
+			runTimes[key(d)] += d.Time.Sub(startTimes[key(d)])
+
+			byPackage[d.Package] = append(byPackage[d.Package], TestRun{
+				Action:  d.Action,
+				Package: d.Package,
+				Test:    d.Test,
+				Output:  outputs[key(d)],
+				RunTime: runTimes[key(d)],
+			})
+		}
+	}
+
+	return byPackage
+}
+
+func printSummary(packages []string, byPackage map[string][]TestRun) {
+	fmt.Printf("## Package Summary\n\n")
+
+	// output table-of-contents
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		if len(tests) == 1 {
+			// package-only
+			continue
+		}
+
+		totalRuntime := time.Duration(0)
+		for _, t := range tests[1:] {
+			totalRuntime += t.RunTime
+		}
+
+		overallOutcome := actionSymbol(tests[0])
+		fmt.Printf("* %s `%s` (runtime %s)\n", overallOutcome, pkg, totalRuntime)
+	}
+
+	fmt.Println()
+}
+
+var maxOutputLines = 300
+
+func printDetails(packages []string, byPackage map[string][]TestRun) {
+	fmt.Printf("## Failed Test Details\n\n")
+
+	for _, pkg := range packages {
+		tests := byPackage[pkg]
+		// check package-level indicator, which will be first ("" test name):
+		if tests[0].Action != "fail" {
+			continue // no failed tests, skip
+		}
+
+		// package name as header
+		fmt.Printf("### `%s`\n\n", pkg)
+
+		// Output info on stderr
+		fmt.Fprintf(os.Stderr, "Package failed: %s\n", pkg)
+
+		for _, test := range tests[1:] {
+			// only printing failed tests
+			if test.Action == "fail" {
+				fmt.Printf("#### `%s`\n", test.Test)
+				fmt.Printf("Failed in %s\n:", test.RunTime)
+
+				trimmedOutput, output := escapeOutput(test.Output)
+				summary := "Test output"
+				if trimmedOutput {
+					summary += fmt.Sprintf(" (trimmed to last %d lines) — full details available in log", maxOutputLines)
+				}
+
+				fmt.Printf("<details><summary>%s</summary><pre>%s</pre></details>\n\n", summary, output)
+
+				// Output info on stderr, so that test failure isn’t silent on console
+				// when running `task ci`, and that full logs are available if they get trimmed
+				fmt.Fprintf(os.Stderr, "- Test failed: %s\n", test.Test)
+				fmt.Fprintln(os.Stderr, "=== TEST OUTPUT ===")
+				for _, outputLine := range test.Output {
+					fmt.Fprint(os.Stderr, outputLine) // note that line already has newline attached
+				}
+				fmt.Fprintln(os.Stderr, "=== END TEST OUTPUT ===")
+			}
+		}
+
+		fmt.Println()
```

Comment:
Suggest writing `None` if there were no failed tests found, so that we don't have an orphan heading.

Reply:
Removed this section entirely if there are no failures.